core/admin/ViewableAdmin.py

```diff
@@ -0,0 +1,144 @@
+from functools import update_wrapper
+
+from django.contrib.admin.options import csrf_protect_m, IncorrectLookupParameters
+from django.core.exceptions import PermissionDenied
+from django.http import HttpResponseRedirect
+from django.template.response import TemplateResponse, SimpleTemplateResponse
+from django.urls import path
+from django.contrib.admin import ModelAdmin
+from django.contrib.auth import get_permission_codename
+from django.utils.translation import gettext as _, ngettext
+
+from core.views.ViewList import RestrictedViewList
+from core.views.common import ModelDetailView
+
+
+class ViewableModelAdmin(ModelAdmin):
+    """
+    A model admin that has an additional view_<model> permission, that allows viewing without editing
+    """
+
+    list_view = RestrictedViewList
+    detail_view_class = ModelDetailView
+
+    def get_detail_view(self):
+        """Get the detail view as a view (function) wrapped to automatically check permissions"""
+
+        def wrap(view):
+            def wrapper(*args, **kwargs):
+                return self.admin_site.admin_view(view)(*args, **kwargs)
+            wrapper.model_admin = self
+            return update_wrapper(wrapper, view)
+
+        detail = self.detail_view_class
+        detail.field_sets = self.fieldsets
+        detail_view = detail.as_view()
+        return wrap(detail_view)
+
+    def has_view_permission(self, request):
+        opts = self.opts
+        codename = get_permission_codename('view', opts)
+        return request.user.has_perm("%s.%s" % (opts.app_label, codename))
+
+    def get_model_perms(self, request):
+        """
+        Return a dict of all perms for this model. This dict has the keys
+        ``add``, ``change``, and ``delete`` mapping to the True/False for each
+        of those actions.
+        """
+        return {
+            'add': self.has_add_permission(request),
+            'change': self.has_change_permission(request),
+            'delete': self.has_delete_permission(request),
+            'view': self.has_view_permission(request)
+        }
+
+    def get_changelist(self, request, **kwargs):
+        return self.list_view
+
+    @csrf_protect_m
+    def viewlist_view(self, request):
+        """Same as the change list, but no actions and does not accept post"""
+
+        from django.contrib.admin.views.main import ERROR_FLAG
```

Comment:
Why not move this import to the top?

Reply:
I copied this from the overridden method. good point though